### PR TITLE
[SPARK-44059] Add analyzer support of named arguments for built-in functions

### PR DESCRIPTION
**Pull request by learningchess2003**
### What changes were proposed in this pull request?
Add analyzer support for named function arguments. 

### Why are the changes needed?
Part of the project needed for general named function argument support.

### Does this PR introduce _any_ user-facing change?
We added support for named arguments for the ```CountMinSketchAgg``` and ```Mask``` SQL functions.

### How was this patch tested?
A new suite was added for this test called NamedArgumentFunctionSuite.

### CODE REVIEW

**Comment by learningchess2003** [comment]
@dtenedor @cloud-fan @anchovYu @MaxGekk  @allisonwang-db Please take a look when you have time!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 1]
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+  override def build(funcName: String, expressions: Seq[Expression]) : Expression =
+    Explode(expressions(0))
+}
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Separates the elements of array `expr` into multiple rows, or the elements of map `expr` into multiple rows and columns. Unless specified otherwise, uses the default column name `col` for elements of the array or `key` and `value` for the elements of the map.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(10, 20));
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:458]: This is the function doc of the TVF version of `explode`. Shall we show the example with the TVF syntax? `SELECT * FROM _FUNC_(array(10, 20))`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 2]
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+object ExplodeGeneratorBuilder extends ExplodeGeneratorBuilderBase {
+  override def isOuter: Boolean = false
+}
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Separates the elements of array `expr` into multiple rows, or the elements of map `expr` into multiple rows and columns. Unless specified otherwise, uses the default column name `col` for elements of the array or `key` and `value` for the elements of the map.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(10, 20));
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:477]: ditto

**Review by cloud-fan — APPROVED** [review]

**Review by dtenedor — APPROVED** [review]
LGTM 👍

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 3]
```diff
@@ -421,16 +435,60 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+       10
+       20
+      > SELECT _FUNC_(collection => array(10, 20));
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:480]: Can we also add one example for `pos_explode`?

**Review by allisonwang-db — COMMENTED** [review]
Looks good!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/FunctionBuilderBase.scala [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * Represents a parameter of a function expression. Function expressions should use this class
+ * to construct the argument lists returned in [[Builder]]
+ *
+ * @param name     The name of the string.
+ * @param default  The default value of the argument. If the default is none, then that means the
+ *                 argument is required. If no argument is provided, an exception is thrown.
+ */
+case class NamedArgument(
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/FunctionBuilderBase.scala:168]: nit: Can this case class fit on the same line?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 5]
```diff
@@ -75,6 +76,25 @@ import org.apache.spark.unsafe.types.UTF8String
...
+    val otherCharArg = NamedArgument(
+      "otherChar",
+      Some(Literal(Mask.MASKED_IGNORE, StringType)))
+    val functionSignature: FunctionSignature = FunctionSignature(Seq(
+      strArg, upperCharArg, lowerCharArg, digitCharArg, otherCharArg))
+    Some(functionSignature)
+  }
+
+  override def build(funcName: String, expressions: Seq[Expression]): Expression = {
+    new Mask(expressions(0), expressions(1), expressions(2), expressions(3), expressions(4))
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:94]: nit: Could you still add an assertion on the size of the expressions? This way `build` here as an independent function is more complete itself.

File: common/utils/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -1956,7 +1974,13 @@
       "Not allowed to implement multiple UDF interfaces, UDF class <className>."
     ]
   },
-  "NAMED_ARGUMENTS_SUPPORT_DISABLED" : {
+  "NAMED_PARAMETERS_NOT_SUPPORTED" : {
```
Comment (anchovYu) [inline common/utils/src/main/resources/error/error-classes.json:1977]: nit: The object is named as `object NamedArgumentsSupport`. Could you unify the naming on the 'parameters' and 'arguments'?

Comment (anchovYu) [reply common/utils/src/main/resources/error/error-classes.json:1977]: nvm, it seems reasonable for the naming of `NamedArgumentsSupport ` but `NAMED_PARAMETERS_NOT_SUPPORTED`.

**Review by anchovYu — APPROVED** [review]
lgtm!

File: common/utils/src/main/resources/error/error-classes.json (hunk 6 above)
Comment (learningchess2003) [reply common/utils/src/main/resources/error/error-classes.json:1977]: Serge was actually the one who recommended the name change. It should be fine.

Comment (anchovYu) [reply common/utils/src/main/resources/error/error-classes.json:1977]: Right, I agree with the change in the error class to parameter, but I was thinking if similar changes should be applied on `NamedArgumentsSupport`, e.g. `NamedParametersSupport`.

Comment (learningchess2003) [reply common/utils/src/main/resources/error/error-classes.json:1977]: Okay, that makes sense.

**Comment by learningchess2003** [comment]
@cloud-fan Thanks for all the comments! I've addressed comments from the other reviewers. All of the tests passed except a flaky which I might rerun to see if we can pass it. Otherwise, I've already seen an instance where we had an all green.  If you think this PR is ready, you can merge it.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5! (half of this feature is already in 3.5)

**Pull request closed without merging**